### ISSUE TITLE
Canvas-LMS on Codespace

### ISSUE THREAD

**Issue by mike32jacobs** [rubygems/rubygems#6092]
--- ERROR REPORT TEMPLATE -------------------------------------------------------

```
Errno::EACCES: Permission denied @ dir_s_mkdir - /usr/local/rvm/gems/ruby-3.1.3/extensions/x86_64-linux/2.7.0/racc-1.6.0
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:250:in `mkdir'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:250:in `fu_mkdir'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:228:in `block (2 levels) in mkdir_p'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:226:in `reverse_each'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:226:in `block in mkdir_p'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:211:in `each'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:211:in `mkdir_p'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:207:in `write_gem_make_out'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:135:in `build_error'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:169:in `rescue in build_extension'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:158:in `build_extension'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:195:in `block in build_extensions'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:192:in `each'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:192:in `build_extensions'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/installer.rb:853:in `build_extensions'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/rubygems_gem_installer.rb:71:in `build_extensions'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/rubygems_gem_installer.rb:28:in `install'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/source/rubygems.rb:204:in `install'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/installer/gem_installer.rb:54:in `install'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/installer/parallel_installer.rb:186:in `do_install'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/installer/parallel_installer.rb:177:in `block in worker_pool'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/worker.rb:62:in `apply_func'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/worker.rb:57:in `block in process_queue'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/worker.rb:54:in `loop'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/worker.rb:54:in `process_queue'
  /usr/local/rvm/gems/ruby-3.1.3/gems/bundler-2.2.33/lib/bundler/worker.rb:91:in `block (2 levels) in create_threads'
```

## Environment

```
Bundler             2.2.33
  Platforms         ruby, x86_64-linux
Ruby                2.7.6p219 (2022-04-12 revision c9c2245c0a25176072e02db9254f0e0c84c805cd) [x86_64-linux]
  Full Path         /opt/ruby/2.7.6/bin/ruby
  Config Dir        /opt/ruby/2.7.6/etc
RubyGems            3.3.10
  Gem Home          /usr/local/rvm/gems/ruby-3.1.3
  Gem Path          /usr/local/rvm/gems/ruby-3.1.3:/usr/local/rvm/rubies/ruby-3.1.3/lib/ruby/gems/3.1.0
  User Home         /home/codespace
  User Path         /home/codespace/.local/share/gem/ruby/2.7.0
  Bin Dir           /usr/local/rvm/gems/ruby-3.1.3/bin
OpenSSL             
  Compiled          OpenSSL 1.1.1f  31 Mar 2020
  Loaded            OpenSSL 1.1.1f  31 Mar 2020
  Cert File         /usr/lib/ssl/cert.pem
  Cert Dir          /usr/lib/ssl/certs
Tools               
  Git               2.25.1
  RVM               1.29.12 (latest)
  rbenv             rbenv 1.2.0
  chruby            not installed
  rubygems-bundler  (1.4.5)
```

## Bundler Build Metadata

```
Built At          2021-12-07
Git SHA           9b5e2a350b
Released Version  true
```

### Gemfile.rails61.lock

```
<No /workspaces/canvas-lms/Gemfile.rails61.lock found>
```

--- TEMPLATE END ----------------------------------------------------------------

Unfortunately, an unexpected error occurred, and Bundler cannot continue.

First, try this link to see if there are any existing issue reports for this error:
https://github.com/rubygems/rubygems/search?q=Permission+denied+%40+dir_s_mkdir+-+%2Fusr%2Flocal%2Frvm%2Fgems%2Fruby-3.1.3%2Fextensions%2Fx86_64-linux%2F2.7.0%2Fracc-1.6.0&type=Issues

If there aren't any reports for this error yet, please fill in the new issue form located at https://github.com/rubygems/rubygems/issues/new?labels=Bundler&template=bundler-related-issue.md, and copy and paste the report template above in there.
@mike32jacobs ➜ /workspaces/canvas-lms (master ✗) $ sudo chown -R codespace:codespace /var/run/postgresql/
@mike32jacobs ➜ /workspaces/canvas-lms (master ✗) $ export PGHOST=localhost
@mike32jacobs ➜ /workspaces/canvas-lms (master ✗) $ /usr/lib/postgresql/12/bin/initdb ~/postgresql-data/ -E utf8
The files belonging to this database system will be owned by user "codespace".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.UTF-8".
The default text search configuration will be set to "english".

Data page checksums are disabled.

creating directory /home/codespace/postgresql-data ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... Etc/UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

initdb: warning: enabling "trust" authentication for local connections
You can change this by editing pg_hba.conf or using the option -A, or
--auth-local and --auth-host, the next time you run initdb.

Success. You can now start the database server using:

    /usr/lib/postgresql/12/bin/pg_ctl -D /home/codespace/postgresql-data/ -l logfile start

@mike32jacobs ➜ /workspaces/canvas-lms (master ✗) $ /usr/lib/postgresql/12/bin/pg_ctl -D ~/postgresql-data/ -l ~/postgresql-data/server.log start
waiting for server to start.... done
server started
@mike32jacobs ➜ /workspaces/canvas-lms (master ✗) $ /usr/lib/postgresql/12/bin/createdb canvas_development
@mike32jacobs ➜ /workspaces/canvas-lms (master ✗) $ bundle exec rails canvas:compile_assets
bundler: command not found: rails
Install missing gem executables with `bundle install`
@mike32jacobs ➜ /workspaces/canvas-lms (master ✗) $ bundle install
Fetching https://github.com/kyler-instructure/rspecq.git
Fetching https://github.com/erickguan/ffi-icu.git
Fetching https://github.com/rda1902/spring-commands-rubocop.git
Fetching https://github.com/twitter/thrift_client.git
Fetching https://github.com/kreynolds/cassandra-cql.git
Fetching https://github.com/instructure/week-of-month.git
Fetching https://github.com/robotdana/json_schemer.git
Fetching https://github.com/rails-api/active_model_serializers.git
Fetching gem metadata from https://rubygems.org/.......
Resolving dependencies..............................
Ignoring debase-0.2.5.beta2 because its extensions are not built. Try: gem pristine debase --version 0.2.5.beta2
Ignoring debug-1.6.3 because its extensions are not built. Try: gem pristine debug --version 1.6.3
Ignoring eventmachine-1.2.7 because its extensions are not built. Try: gem pristine eventmachine --version 1.2.7
Ignoring executable-hooks-1.6.1 because its extensions are not built. Try: gem pristine executable-hooks --version 1.6.1
Ignoring gem-wrappers-1.4.0 because its extensions are not built. Try: gem pristine gem-wrappers --version 1.4.0
Ignoring http_parser.rb-0.8.0 because its extensions are not built. Try: gem pristine http_parser.rb --version 0.8.0
Ignoring rbs-2.7.0 because its extensions are not built. Try: gem pristine rbs --version 2.7.0
Ignoring sassc-2.4.0 because its extensions are not built. Try: gem pristine sassc --version 2.4.0
Error loading RubyGems plugin "/usr/local/rvm/rubies/ruby-3.1.3/lib/ruby/gems/3.1.0/plugins/gem-wrappers_plugin.rb": cannot load such file -- gem-wrappers (LoadError)
Error loading RubyGems plugin "/usr/local/rvm/rubies/ruby-3.1.3/lib/ruby/gems/3.1.0/plugins/executable-hooks_plugin.rb": cannot load such file -- executable-hooks/wrapper (LoadError)
Using rake 13.0.3
Using Ascii85 1.1.0
Using minitest 5.16.3
Using zeitwerk 2.6.6
Using concurrent-ruby 1.1.10
Using mime-types-data 3.2021.0901
Using multi_xml 0.6.0
Using builder 3.2.4
Using erubi 1.11.0
Fetching racc 1.6.0
Using crass 1.0.6
Using rack 2.2.3
Fetching nio4r 2.5.8
Installing racc 1.6.0 with native extensions
Installing nio4r 2.5.8 with native extensions
Using websocket-extensions 0.1.5
Using marcel 1.0.2
Using mini_mime 1.1.2
Using bundler 2.3.10
Using method_source 1.0.0
Using thor 1.2.1
Using active_record_query_trace 1.8
Using public_suffix 4.0.7
Using aes_key_wrap 1.1.0
Using afm 0.2.2
Using google-protobuf 3.21.10 (x86_64-linux)
Using graphql 1.12.14
Using ast 2.4.2
Using simple_uuid 0.4.0
Fetching thrift 0.9.3.0
Installing thrift 0.9.3.0 with native extensions
Using deep_merge 1.2.2
Using multipart-post 2.1.1
Using folio-pagination 0.0.12
Fetching json 2.6.2
Installing json 2.6.2 with native extensions
Using will_paginate 3.3.0
Using dogstatsd-ruby 4.8.2
Using statsd-ruby 1.4.0
Using parallel 1.22.0
Using autoextend 1.0.0 from source at `gems`
Using extlib 0.9.16
Using multi_json 1.15.0
Using awesome_print 1.9.2
Using aws-eventstream 1.2.0
Using aws-partitions 1.393.0
Using jmespath 1.6.2
Using matrix 0.4.2
Using regexp_parser 1.8.2
Using dumb_delegator 1.0.0
Using childprocess 3.0.0
Using rexml 3.2.5
Using rubyzip 2.3.0
Using thread_safe 0.3.6
Using ice_nine 0.11.2
Using barby 0.6.8
Fetching bcrypt 3.1.16
Installing bcrypt 3.1.16 with native extensions
Using bindata 2.4.14
Fetching bluecloth 2.2.0
Installing bluecloth 2.2.0 with native extensions
Fetching msgpack 1.6.0
Installing msgpack 1.6.0 with native extensions
Using brakeman 5.4.0
Fetching brotli 0.4.0
Using browser 5.1.0
Fetching rbtree 0.4.5
Installing rbtree 0.4.5 with native extensions
Using set 1.0.3
Fetching byebug 11.1.3
Installing byebug 11.1.3 with native extensions
Installing brotli 0.4.0 with native extensions
Fetching digest-murmurhash 1.1.1
Installing digest-murmurhash 1.1.1 with native extensions
Using redis 4.1.4
Using canvas_color 0.0.1 from source at `gems`
Using canvas_crummy 0.0.1 from source at `gems`
Fetching debug_inspector 1.1.0
Installing debug_inspector 1.1.0 with native extensions
Using raabro 1.4.0
Using swearjar 1.4.0
Using canvas_sort 1.0.0 from source at `gems`
Using canvas_mimetype_fu 0.0.1 from source at `gems`
Fetching pg 1.2.3
Installing pg 1.2.3 with native extensions
Fetching idn-ruby 0.1.5
Installing idn-ruby 0.1.5 with native extensions
Fetching unf_ext 0.0.8.2
Installing unf_ext 0.0.8.2 with native extensions
Using chunky_png 1.3.14
Using colored 1.2
Using colorize 0.8.1
Using rchardet 1.8.0
Fetching sqlite3 1.4.2
Installing sqlite3 1.4.2 with native extensions
Using database_cleaner-core 2.0.1
Using rspec-support 3.12.0
Using diff-lcs 1.5.0
Using debase-ruby_core_source 0.10.18
Using dig_rb 1.0.1
Using docile 1.1.5
Using dotenv 2.7.5
Using mustache 1.1.1
Using pygments.rb 2.2.0
Using escape_code 0.2
Fetching redcarpet 3.5.0
Fetching ffi 1.14.2
Installing redcarpet 3.5.0 with native extensions
Using filelock 1.1.1
Using find_a_port 1.0.1
Using git-version-bump 0.17.2
Using globby 0.1.2
Using retriable 1.4.1
Using jwt 2.5.0
Using promise.rb 0.7.4
Using hana 1.3.7
Using hashdiff 1.0.1
Using hashery 2.1.2
Using headless 2.3.1
Using highline 2.0.3
Using http-accept 1.7.0
Using sexp_processor 4.16.1
Using ya2yaml 0.31
Using utf8_cleaner 0.0.1 from source at `gems`
Using ice_cube 0.16.4
Using simple_oauth 0.3.1
Fetching io-console 0.5.11
Installing io-console 0.5.11 with native extensions
Installing ffi 1.14.2 with native extensions
Using iso8601 0.13.0
Using jira_ref_parser 1.0.1
Using uri_template 0.7.0
Using link_header 0.0.8
Using oauth 0.5.4
Using luminosity_contrast 0.2.1
Using mini_magick 4.11.0
Fetching rdiscount 2.2.7
Installing rdiscount 2.2.7 with native extensions
Using net-ldap 0.16.3
Using netrc 0.11.0
Fetching oj 3.10.16
Installing oj 3.10.16 with native extensions
Using open4 1.3.4
Using randexp 0.1.7
Using sync 0.5.0
Using webrick 1.7.0
Using table_print 1.5.7
Using pdf-core 0.9.0
Using ruby-rc4 0.1.5
Using ttfunk 1.7.0
Using rainbow 3.0.0
Using soap4r-ruby1.9 2.0.5
Using ritex 1.0.1
Using rotp 6.2.0
Using rqrcode_core 0.2.0
Using ruby-progressbar 1.11.0
Using unicode-display_width 2.1.0
Using ruby-debug-ide 0.7.3
Using ruby-rtf 0.0.5
Using ruby2_keywords 0.0.3
Using simplecov-html 0.12.3
Using simplecov_json_formatter 0.1.4
Using spring 4.0.0
Fetching stackprof 0.2.23
Installing stackprof 0.2.23 with native extensions
Using stringify_ids 1.0.0 from source at `gems`
Using testrail_client 0.0.1
Using timecop 0.9.1
Using wcag_color_contrast 0.1.0
Using week_of_month 1.2.5 from https://github.com/instructure/week-of-month.git (at b301363@b301363)
Using yard 0.9.25
Using i18n 1.8.5
Using tzinfo 2.0.4
Using sentry-ruby-core 5.1.0
Using mime-types 3.3.1
Using rack-test 1.1.0
Using sprockets 4.1.1
Using request_store 1.5.1
Ignoring debase-0.2.5.beta2 because its extensions are not built. Try: gem pristine debase --version 0.2.5.beta2
Ignoring debug-1.6.3 because its extensions are not built. Try: gem pristine debug --version 1.6.3
Ignoring eventmachine-1.2.7 because its extensions are not built. Try: gem pristine eventmachine --version 1.2.7
Ignoring executable-hooks-1.6.1 because its extensions are not built. Try: gem pristine executable-hooks --version 1.6.1
Ignoring gem-wrappers-1.4.0 because its extensions are not built. Try: gem pristine gem-wrappers --version 1.4.0
Ignoring http_parser.rb-0.8.0 because its extensions are not built. Try: gem pristine http_parser.rb --version 0.8.0
Ignoring rbs-2.7.0 because its extensions are not built. Try: gem pristine rbs --version 2.7.0
Ignoring sassc-2.4.0 because its extensions are not built. Try: gem pristine sassc --version 2.4.0
Error loading RubyGems plugin "/usr/local/rvm/rubies/ruby-3.1.3/lib/ruby/gems/3.1.0/plugins/gem-wrappers_plugin.rb": cannot load such file -- gem-wrappers (LoadError)
Error loading RubyGems plugin "/usr/local/rvm/rubies/ruby-3.1.3/lib/ruby/gems/3.1.0/plugins/executable-hooks_plugin.rb": cannot load such file -- executable-hooks/wrapper (LoadError)
Using rake 13.0.3
Using Ascii85 1.1.0
Using concurrent-ruby 1.1.10
Using minitest 5.16.3
Using zeitwerk 2.6.6
Using mime-types-data 3.2021.0901
Using multi_xml 0.6.0
Using builder 3.2.4
Using erubi 1.11.0
Fetching racc 1.6.0
Using crass 1.0.6
Using rack 2.2.3
Installing racc 1.6.0 with native extensions
Fetching nio4r 2.5.8
Installing nio4r 2.5.8 with native extensions
Using websocket-extensions 0.1.5
Using marcel 1.0.2
Using mini_mime 1.1.2
Using bundler 2.3.10
Using method_source 1.0.0
Using thor 1.2.1
Using active_record_query_trace 1.8
Using public_suffix 4.0.7
Using aes_key_wrap 1.1.0
Using afm 0.2.2
Using google-protobuf 3.21.10 (x86_64-linux)
Using graphql 1.12.14
Using ast 2.4.2
Using simple_uuid 0.4.0
Fetching thrift 0.9.3.0
Installing thrift 0.9.3.0 with native extensions
Using deep_merge 1.2.2
Using multipart-post 2.1.1
Using folio-pagination 0.0.12
Fetching json 2.6.2
Installing json 2.6.2 with native extensions
Using will_paginate 3.3.0
Using dogstatsd-ruby 4.8.2
Using statsd-ruby 1.4.0
Using parallel 1.22.0
Using autoextend 1.0.0 from source at `gems`
Using extlib 0.9.16
Using multi_json 1.15.0
Using awesome_print 1.9.2
Using aws-eventstream 1.2.0
Using aws-partitions 1.393.0
Using jmespath 1.6.2
Using matrix 0.4.2
Using regexp_parser 1.8.2
Using dumb_delegator 1.0.0
Using childprocess 3.0.0
Using rexml 3.2.5
Using rubyzip 2.3.0
Using thread_safe 0.3.6
Using ice_nine 0.11.2
Using barby 0.6.8
Fetching bcrypt 3.1.16
Installing bcrypt 3.1.16 with native extensions
Using bindata 2.4.14
Fetching bluecloth 2.2.0
Installing bluecloth 2.2.0 with native extensions
Fetching msgpack 1.6.0
Installing msgpack 1.6.0 with native extensions
Using brakeman 5.4.0
Fetching brotli 0.4.0
Using browser 5.1.0
Fetching rbtree 0.4.5
Installing brotli 0.4.0 with native extensions
Installing rbtree 0.4.5 with native extensions
Using set 1.0.3
Fetching byebug 11.1.3
Installing byebug 11.1.3 with native extensions
Fetching digest-murmurhash 1.1.1
Installing digest-murmurhash 1.1.1 with native extensions
Using redis 4.1.4
Using canvas_color 0.0.1 from source at `gems`
Using canvas_crummy 0.0.1 from source at `gems`
Fetching debug_inspector 1.1.0
Installing debug_inspector 1.1.0 with native extensions
Using raabro 1.4.0
Using swearjar 1.4.0
Using canvas_mimetype_fu 0.0.1 from source at `gems`
Using canvas_sort 1.0.0 from source at `gems`
Fetching pg 1.2.3
Fetching idn-ruby 0.1.5
Installing idn-ruby 0.1.5 with native extensions
Installing pg 1.2.3 with native extensions
Fetching unf_ext 0.0.8.2
Installing unf_ext 0.0.8.2 with native extensions
Using chunky_png 1.3.14
Using colored 1.2
Using colorize 0.8.1
Using rchardet 1.8.0
Fetching sqlite3 1.4.2
Using database_cleaner-core 2.0.1
Using rspec-support 3.12.0
Using diff-lcs 1.5.0
Using debase-ruby_core_source 0.10.18
Using dig_rb 1.0.1
Using docile 1.1.5
Using dotenv 2.7.5
Using mustache 1.1.1
Installing sqlite3 1.4.2 with native extensions
Using pygments.rb 2.2.0
Fetching redcarpet 3.5.0
Installing redcarpet 3.5.0 with native extensions
Using escape_code 0.2
Fetching ffi 1.14.2
Using filelock 1.1.1
Using find_a_port 1.0.1
Using git-version-bump 0.17.2
Using globby 0.1.2
Using retriable 1.4.1
Using jwt 2.5.0
Using promise.rb 0.7.4
Using hana 1.3.7
Using hashdiff 1.0.1
Using hashery 2.1.2
Using headless 2.3.1
Using highline 2.0.3
Using http-accept 1.7.0
Using sexp_processor 4.16.1
Using ya2yaml 0.31
Using utf8_cleaner 0.0.1 from source at `gems`
Using ice_cube 0.16.4
Using simple_oauth 0.3.1
Fetching io-console 0.5.11
Installing io-console 0.5.11 with native extensions
Using iso8601 0.13.0
Using jira_ref_parser 1.0.1
Using uri_template 0.7.0
Using link_header 0.0.8
Using oauth 0.5.4
Using luminosity_contrast 0.2.1
Using mini_magick 4.11.0
Fetching rdiscount 2.2.7
Installing ffi 1.14.2 with native extensions
Installing rdiscount 2.2.7 with native extensions
Using net-ldap 0.16.3
Using netrc 0.11.0
Fetching oj 3.10.16
Installing oj 3.10.16 with native extensions
Using open4 1.3.4
Using randexp 0.1.7
Using sync 0.5.0
Using webrick 1.7.0
Using table_print 1.5.7
Using pdf-core 0.9.0
Using ruby-rc4 0.1.5
Using ttfunk 1.7.0
Using rainbow 3.0.0
Using soap4r-ruby1.9 2.0.5
Using ritex 1.0.1
Using rotp 6.2.0
Using rqrcode_core 0.2.0
Using ruby-progressbar 1.11.0
Using unicode-display_width 2.1.0
Using ruby-debug-ide 0.7.3
Using ruby-rtf 0.0.5
Using ruby2_keywords 0.0.3
Using simplecov-html 0.12.3
Using simplecov_json_formatter 0.1.4
Using spring 4.0.0
Fetching stackprof 0.2.23
Installing stackprof 0.2.23 with native extensions
Using stringify_ids 1.0.0 from source at `gems`
Using testrail_client 0.0.1
Using timecop 0.9.1
Using wcag_color_contrast 0.1.0
Using week_of_month 1.2.5 from https://github.com/instructure/week-of-month.git (at b301363@b301363)
Using yard 0.9.25
Using i18n 1.8.5
Using tzinfo 2.0.4
Using sentry-ruby-core 5.1.0
Using mime-types 3.3.1
Using rack-test 1.1.0
Using sprockets 4.1.1
Using request_store 1.5.1
--- ERROR REPORT TEMPLATE -------------------------------------------------------

```
Errno::EACCES: Permission denied @ dir_s_mkdir - /usr/local/rvm/gems/ruby-3.1.3/extensions/x86_64-linux/2.7.0/racc-1.6.0
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:250:in `mkdir'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:250:in `fu_mkdir'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:228:in `block (2 levels) in mkdir_p'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:226:in `reverse_each'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:226:in `block in mkdir_p'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:211:in `each'
  /opt/ruby/2.7.6/lib/ruby/2.7.0/fileutils.rb:211:in `mkdir_p'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:207:in `write_gem_make_out'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:135:in `build_error'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:169:in `rescue in build_extension'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:158:in `build_extension'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:195:in `block in build_extensions'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:192:in `each'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/ext/builder.rb:192:in `build_extensions'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/rubygems/installer.rb:853:in `build_extensions'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/rubygems_gem_installer.rb:71:in `build_extensions'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/rubygems_gem_installer.rb:28:in `install'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/source/rubygems.rb:204:in `install'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/installer/gem_installer.rb:54:in `install'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/installer/gem_installer.rb:16:in `install_from_spec'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/installer/parallel_installer.rb:186:in `do_install'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/installer/parallel_installer.rb:177:in `block in worker_pool'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/worker.rb:62:in `apply_func'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/worker.rb:57:in `block in process_queue'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/worker.rb:54:in `loop'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/worker.rb:54:in `process_queue'
  /opt/ruby/2.7.6/lib/ruby/site_ruby/2.7.0/bundler/worker.rb:91:in `block (2 levels) in create_threads'
```

## Environment

```
Bundler             2.3.10
  Platforms         ruby, x86_64-linux
Ruby                2.7.6p219 (2022-04-12 revision c9c2245c0a25176072e02db9254f0e0c84c805cd) [x86_64-linux]
  Full Path         /opt/ruby/2.7.6/bin/ruby
  Config Dir        /opt/ruby/2.7.6/etc
RubyGems            3.3.10
  Gem Home          /usr/local/rvm/gems/ruby-3.1.3
  Gem Path          /usr/local/rvm/gems/ruby-3.1.3:/usr/local/rvm/rubies/ruby-3.1.3/lib/ruby/gems/3.1.0
  User Home         /home/codespace
  User Path         /home/codespace/.local/share/gem/ruby/2.7.0
  Bin Dir           /usr/local/rvm/gems/ruby-3.1.3/bin
OpenSSL             
  Compiled          OpenSSL 1.1.1f  31 Mar 2020
  Loaded            OpenSSL 1.1.1f  31 Mar 2020
  Cert File         /usr/lib/ssl/cert.pem
  Cert Dir          /usr/lib/ssl/certs
Tools               
  Git               2.25.1
  RVM               1.29.12 (latest)
  rbenv             rbenv 1.2.0
  chruby            not installed
  rubygems-bundler  (1.4.5)
```

## Bundler Build Metadata

```
Built At          2022-03-23
Git SHA           4bbb70e7de
Released Version  true
```

### Gemfile.rails61.lock

```
<No /workspaces/canvas-lms/Gemfile.rails61.lock found>
```

--- TEMPLATE END ---------------------------------------------------------------

**Comment by deivid-rodriguez** [comment]
Hi, it seems that the Codespace is using a global installation of Ruby at `/opt/ruby/2.7.6/`, and a RVM-managed GEM_HOME at `/usr/local/rvm/gems/ruby-3.1.3`.

Not sure why this is setup like that, it seems messed up. I recommend to uninstall RVM and try again.

It doesn't look like an issue in RubyGems or Bundler.

**Comment by mike32jacobs** [comment]
Thanks. I will give it a try.

**Comment by isl-dbouman** [comment]
@mike32jacobs can you please share how you uninstalled RVM? I'm getting the exact same issue here, but I can't figure out how to fix this.

**Comment by mike32jacobs** [comment]
@isl-dbouman Unfortunately I could not get this to work. I tried various installation setups. 

- [Docker Codespaces](https://groups.google.com/g/canvas-lms-users/c/RZaPIggmC9g/m/Mfsdg-XHCgAJ)
- [Choosing a server?](https://groups.google.com/g/canvas-lms-users/c/QDYrIQfsp58/m/aeEJ8yjICgAJ)
- [Installing Canvas using ./script/docker_dev_setup.sh but having an issue with dory #60](https://github.com/instructure/canvas-lms/issues) 
- [/script/docker_dev_setup.sh on linux #2095](https://github.com/instructure/canvas-lms/issues/2095#issuecomment-1303603201)

The only one that I got functioning was from [Bitnami](https://bitnami.com/stack/canvaslms/cloud).  Unfortunately it will be deprecated. 
If you do get it figured out, I would love to know. I think I need to learn a lot more about dev before I can get this going. I expected this installation process to be rather simple.

**Comment by isl-dbouman** [comment]
@mike32jacobs looks like you and I have tried very similar things, been trying unsuccessfully to get this working for the past 2 days.

@deivid-rodriguez any chance you could give us some insight here? There must be a step missing in the guide because we can't seem to figure out how to get this working.

**Comment by deivid-rodriguez** [comment]
Hei! Could you reduce this issue to a minimum for me? Ideally create a sample repository and right some commands I need to run on my machine to see the issue for myself. I know I can try follow those tutorials, but the more you can simplify this for me, the more chances I can get to having a look quicker.

**Comment by deivid-rodriguez** [comment]
I'm going to close this ticket since we didn't get further feedback. Happy to have a closer look if you can help me out though.

**Comment by isl-dbouman** [comment]
@deivid-rodriguez I apologize for the delay, here are the minimum steps to reproduce this issue:

1. Fork https://github.com/instructure/canvas-lms
2. Switch to “prod” branch
3. Create codespace on prod
4. Following this guide - https://github.com/instructure/canvas-lms/wiki/Quick-Start. Run the following commands:
```
sudo apt-get update
sudo apt-get -y install ruby ruby-dev postgresql-12 zlib1g-dev libxml2-dev libsqlite3-dev libpq-dev libxmlsec1-dev curl build-essential
sudo gem install bundle
sudo gem install bundler:2.2.33
sudo gem install nokogumbo scrypt sanitize ruby-debug-ide
sudo chown -R codespace:codespace /workspaces/canvas-lms/
sudo chown -R codespace:codespace /var/lib/gems/2.7.0/
bundle _2.2.33_ install
```
Observe error mentioned above in this issue.

**Comment by deivid-rodriguez** [comment]
I didn't have time yet to try reproduce this issue, but I did find where this setup comes from: https://github.com/devcontainers/features/blob/dbf069a5db328e0413300094a0dc37f520000375/src/ruby/install.sh.

I have no idea why this container installs both Rbenv and RVM at the same time, it seems calling for trouble to me 🤷‍♂️. But I don't think that's the issue here.

This part does seem more suspicious:

https://github.com/devcontainers/features/blob/dbf069a5db328e0413300094a0dc37f520000375/src/ruby/install.sh#L283-L286

Can you show the permissions of `/usr/local/rvm/gems/ruby-3.1.3/extensions/x86_64-linux/2.7.0/racc-1.6.0` and each parent folder after the error happens?

**Comment by isl-dbouman** [comment]
Thank you for taking a look! Here is a look at the permissions, please let me know if this is enough or if you need to see more parent folders.

![Screen Shot 2022-12-29 at 2 10 10 PM](https://user-images.githubusercontent.com/97130620/209998310-25d000c0-f67e-43b7-bc96-456866164374.png)
![Screen Shot 2022-12-29 at 2 09 37 PM](https://user-images.githubusercontent.com/97130620/209998313-bbd81fe0-f163-490f-89aa-39a7922d4892.png)
![Screen Shot 2022-12-29 at 2 09 04 PM](https://user-images.githubusercontent.com/97130620/209998315-90b9f7a2-c168-45ed-9386-950d9cec5409.png)
![Screen Shot 2022-12-29 at 2 08 30 PM](https://user-images.githubusercontent.com/97130620/209998316-8d04ae53-191b-4a49-800e-c7498f36a26b.png)
![Screen Shot 2022-12-29 at 2 06 32 PM](https://user-images.githubusercontent.com/97130620/209998319-0ff60cb0-dce3-4451-b961-779fe42b58f7.png)

**Comment by deivid-rodriguez** [comment]
I think the following is the explanation of what's going on:

### TL;DR

That wiki page tutorial is wrong, please submit an issue to Canvas-LMS so that they fix their "Quick Start" instructions. I think the long explanation below should be useful for Canvas-LMS maintainers to fix their instructions.

### Long explanation

I don't see any codespace configuration in that repository. When there's no codespace configuration, GitHub will use a "universal image", which is defined [here](https://github.com/devcontainers/images/blob/e0d05d84fd4d7b45e20688daee299fc8adbc1911/src/universal/README.md). That image includes Ruby as a feature [here](https://github.com/devcontainers/images/blob/e0d05d84fd4d7b45e20688daee299fc8adbc1911/src/universal/.devcontainer/devcontainer.json#L40-L43). That's the install script I linked to before. This means a default codespace includes Ruby 3.1 (and Ruby 3.0 additionally), both installed and managed by RVM.

Quick start instructions start by installing a OS-packaged Ruby (`apt-get install ruby`), which is Ruby 2.7. This should take no effect since RVM is configured, which takes priority. However, then there are several instructions recommending different `sudo gem install` commands. These are all wrong and unnecessary. The are several reasons for this which explain the error reported here:

* `sudo` does not preserve the original `PATH` when invoked, so the binstubs installed by those commands will have a shebang pointing to the OS-packaged ruby (because RVM is no longer in PATH), and that will cause the OS packaged Ruby to be run when those binstubs are executed. In particular, because of this, the `bundle _2.2.33_ install` command will run under Ruby 2.7, but with environment configured for Ruby 3.1. This explains the mix and match between Ruby versions in the backtrace of the error.

* Those `sudo gem install` not only install wrong binstubs, but also corrupt permissions of RVM managed folders, because they will be created with root permissions, and when the low privileges command `bundle _2.2.33_ install` runs, it cannot create any installed files on those folders due to this. This explains the `Errno::EACCES: Permission denied @ dir_s_mkdir - /usr/local/rvm/gems/ruby-3.1.3/extensions/x86_64-linux/2.7.0/racc-1.6.0
` wording in the error reported.

### How to get unblocked

There's several things you can try until you get an answer from Canvas-LMS maintainers, such as:

* Completely ignore all `sudo` commands and use preinstalled Ruby and Bundler. So, run `bundle install` as soon as the codespace is built.
* If Canvas-LMS really needs Ruby 2.7, then use RVM to install and use that Ruby. Head over to https://github.com/rvm/rvm for instructions on how to do this. And again, don't install OS-packaged Ruby through `apt` get and don't run any `gem install` commands with `sudo`.

I hope you found this useful and can get rid of this problem and focus on what you were tying to do. Good luck!

**Comment by ZAKosma** [comment]
Thank you @deivid-rodriguez. This is still an issue today. I'll mention it to Canvas.
